[FSDP] Fix `clip_grad_norm_()` for low prec grads

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90028
* #90027

For PyTorch FSDP, the only way that gradients are in low precision is if `keep_low_precision_grads=True` or if the user turns on AMP. This PR adds tests for the former and improves the documentation for `clip_grad_norm_()`, especially around these non-full-precision cases.

Commit:
[FSDP] Fix `clip_grad_norm_()` for low prec grads

[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix `clip_grad_norm_()` for low prec grads"




[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix `clip_grad_norm_()` for low prec grads"


For PyTorch FSDP, the only way that gradients are in low precision is if `keep_low_precision_grads=True` or if the user turns on AMP. This PR adds tests for the former and improves the documentation for `clip_grad_norm_()`, especially around these non-full-precision cases.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1171,12 +1171,21 @@ def clip_grad_norm_(
-            applied per subset of model parameters.
+        .. note:: If every FSDP instance uses ``NO_SHARD``, meaning that no
+            gradients are sharded across ranks, then you may directly use
+            :func:`torch.nn.utils.clip_grad_norm_`.
```
Can we warn explicitly about this?

awgu replied:
I have it so that if all instances use `NO_SHARD`, then this method returns `torch.nn.utils.clip_grad_norm_()`, so it is equivalent.

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1171,12 +1171,21 @@ def clip_grad_norm_(
+            semantics. For example, if *all* parameters/gradients use a low
+            precision dtype, then the returned dtype will be that low precision
+            dtype, but if there exists at least one parameter/gradient using
+            FP32, then the returned dtype will be FP32.
```
nit: "then the returned norm will have dtype fp32"

rohan-varma reviewed (approved):
Great catch! Do we know why this was not picked up by unittests earlier?

Commit:
Update on "[FSDP] Fix `clip_grad_norm_()` for low prec grads"


For PyTorch FSDP, the only way that gradients are in low precision is if `keep_low_precision_grads=True` or if the user turns on AMP. This PR adds tests for the former and improves the documentation for `clip_grad_norm_()`, especially around these non-full-precision cases.

[ghstack-poisoned]

awgu commented:
> Great catch! Do we know why this was not picked up by unittests earlier?

We did not not test `keep_low_precision_grads=True` previously :/

Commit:
Update on "[FSDP] Fix `clip_grad_norm_()` for low prec grads"


For PyTorch FSDP, the only way that gradients are in low precision is if `keep_low_precision_grads=True` or if the user turns on AMP. This PR adds tests for the former and improves the documentation for `clip_grad_norm_()`, especially around these non-full-precision cases.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Fix `clip_grad_norm_()` for low prec grads"


For PyTorch FSDP, the only way that gradients are in low precision is if `keep_low_precision_grads=True` or if the user turns on AMP. This PR adds tests for the former and improves the documentation for `clip_grad_norm_()`, especially around these non-full-precision cases.

[ghstack-poisoned]

awgu commented:
@pytorchbot rebase -s

Commit:
Update on "[FSDP] Fix `clip_grad_norm_()` for low prec grads"


For PyTorch FSDP, the only way that gradients are in low precision is if `keep_low_precision_grads=True` or if the user turns on AMP. This PR adds tests for the former and improves the documentation for `clip_grad_norm_()`, especially around these non-full-precision cases.

[ghstack-poisoned]
